qiskit/transpiler/preset_passmanagers/plugin.py

```diff
@@ -0,0 +1,256 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2022.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""
+=======================================================================================
+Transpiler Stage Plugin Interface (:mod:`qiskit.transpiler.preset_passmanagers.plugin`)
+=======================================================================================
+
+.. currentmodule:: qiskit.transpiler.preset_passmanagers.plugin
+
+This module defines the plugin interface for providing custom stage
+implementations for the preset pass managers and the :func:`~.transpile`
+function. This enables external Python packages to provide
+:class:`~.PassManager` objects that can be used for each stage.
+
+The plugin interfaces are built using setuptools
+`entry points <https://setuptools.readthedocs.io/en/latest/userguide/entry_point.html>`__
+which enable packages external to Qiskit to advertise they include a transpiler stage.
+
+See :mod:`qiskit.transpiler.passes.synthesis.plugin` for details on how to
+write plugins for synthesis methods which are used by the transpiler.
+
+.. _stage_table:
+
+Plugin Stages
+=============
+
+Currently there are 6 stages in the preset pass managers used by and corresponding entrypoints.
+
+.. list-table:: Stages
+   :header-rows: 1
+
+   * - Stage Name
+     - Entry Point
+     - Reserved Names
+     - Description and expectations
+   * - ``init``
+     - ``qiskit.transpiler.init``
+     - No reserved names
+     - This stage runs first and is typically used for an initial logical optimization also
+       for most ``layout`` and ``routing`` stages this stage is used to translate any gates
+       that operate on more than 2 qubits into gates that operate on 1 or 2
+       qubits only. This is because most layout and routing algorithms are
+       only designed to work with 1 and 2 qubit gates.
+       ``init``
+   * - ``layout``
+     - ``qiskit.transpiler.layout``
+     - ``trivial``, ``dense``, ``noise_adaptive``, ``sabre``
+     - The output from this staged is expected to have the ``layout`` property
+       set field set with a :class:`~.Layout` object. Additionally, the circuit is
+       is typically expected to be embedded so that it expanded to include all
+       qubits and the :class:`~.ApplyLayout` pass is expected to be run to apply the
+       layout. The embedding of the :class:`~.Layout` can be generated with
+       :func:`~.generate_embed_passmanager`.
```

Comment:
No idea why, but in the docs `generate_embed_passmanager` is not clickable.

Reply:
The function might not be indexed by autodoc/autosummary anywhere. We can fix this (probably in a separate pr)